src/components/templates/Camp/hooks/useKeyboardControls.ts

```diff
@@ -0,0 +1,219 @@
+import { useCallback, useEffect, useState } from "react";
+import { FirebaseReducer } from "react-redux-firebase";
+
+import { CampRoomData } from "types/CampRoomData";
+import { CampVenue } from "types/CampVenue";
+import { orderedVenuesSelector } from "utils/selectors";
+import { User } from "types/User";
+import { Matrix, ReactHook } from "types/utility";
+
+import { WithId } from "utils/id";
+import { currentTimeInUnixEpoch } from "utils/time";
+import { openRoomUrl } from "utils/url";
+import { enterRoom } from "utils/useLocationUpdateEffect";
+
+import { useSelector } from "hooks/useSelector";
+
+import { DEFAULT_COLUMNS, DEFAULT_ROWS } from "../components/Map";
+
+import { useKeyPress } from "./useKeyPress";
+
+export const MOVEMENT_INTERVAL = 350;
+
+interface UseKeyboardControlsProps {
+  // From useUser() hook
+  user?: FirebaseReducer.AuthState;
+  profile?: FirebaseReducer.Profile<User>;
+
+  // Passed down from Map component
+  rows: number;
+  columns: number;
+  partygoersBySeat: Matrix<WithId<User>>;
+  isHittingRoom: boolean;
+  setIsHittingRoom: (value: boolean) => void;
+  takeSeat: (row: number | null, column: number | null) => void;
+
+  // Passed down from Camp component
+  venue: CampVenue;
+  selectedRoom?: CampRoomData;
+  setSelectedRoom: (room: CampRoomData | undefined) => void;
+}
+
+interface KeyboardControls {
+  roomEnter: (room: CampRoomData) => void;
+}
+
+export const useKeyboardControls: ReactHook<
+  UseKeyboardControlsProps,
+  KeyboardControls
+> = ({
+  user,
+  profile,
+
+  venue,
+  partygoersBySeat,
+
+  rows,
+  columns,
+  isHittingRoom,
+  setIsHittingRoom,
+  takeSeat,
+
+  selectedRoom,
+  setSelectedRoom,
+}) => {
+  const [keyDown, setKeyDown] = useState(false);
+
+  const venues = useSelector(orderedVenuesSelector);
+
+  const currentPosition = profile?.data?.[venue.id];
+
+  const downPress = useKeyPress("ArrowDown");
+  const upPress = useKeyPress("ArrowUp");
+  const leftPress = useKeyPress("ArrowLeft");
+  const rightPress = useKeyPress("ArrowRight");
+  const enterPress = useKeyPress("Enter");
+
+  const roomEnter = useCallback(
+    (room: CampRoomData) => {
+      const roomVenue = venues?.find((venue) =>
+        room.url.endsWith(`/${venue.id}`)
+      );
+      const venueRoom = roomVenue
+        ? { [roomVenue.name]: currentTimeInUnixEpoch }
+        : {};
+      room &&
+        user &&
+        enterRoom(
+          user,
+          {
+            [`${venue.name}/${room.title}`]: currentTimeInUnixEpoch,
+            ...venueRoom,
+          },
+          profile?.lastSeenIn
+        );
+    },
+    [profile, user, venue, venues]
+  );
+
+  // TODO: this is duplicated logic from handlePotentialRoomClick in Map component, so we can probably simplify
+  // TODO: do we want to filter this? Or just handle 1 room hit
+  const hitRoom = useCallback(
+    (row: number, column: number) => {
+      let isHitting = false;
+      venue.rooms.forEach((room: CampRoomData) => {
```

Comment:
[TODO] see if we can de-duplicate this logic (and in doing so only have to process the rooms once rather than twice. If there are lots of rooms this could make a significant difference.

I was thinking of filtering the `venue.rooms` so that it only contains the list of 'hit' rooms, which in theory would only be one.

The current logic using `forEach` isn't going to stop when it finds a hit, which the rest of the code seems to imply it might want to do. So even if we just change this to use `.find()` or similar, could help:

- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find